frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -0,0 +1,846 @@
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    #[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance")]
+    pub enum Event<T: Config> {
+        /// Account has bonded and staked funds on a smart contract.
+        BondAndStake(T::AccountId, SmartContract<T::AccountId>, BalanceOf<T>),
+        /// Account has unbonded, unstaked and withdrawn funds.
+        UnbondUnstakeAndWithdraw(T::AccountId, SmartContract<T::AccountId>, BalanceOf<T>),
+        /// New contract added for staking, with deposit value
+        NewContract(T::AccountId, SmartContract<T::AccountId>),
+        /// New dapps staking era. Distribute era rewards to contracts
+        NewDappStakingEra(EraIndex),
+        /// The contract's reward have been claimed, by an account, from era, until era
+        ContractClaimed(
+            SmartContract<T::AccountId>,
+            T::AccountId,
+            EraIndex,
+            EraIndex,
+        ),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Can not stake with zero value.
+        StakingWithNoValue,
+        /// Can not stake with value less than minimum staking value
+        InsufficientStakingValue,
+        /// Number of stakers per contract exceeded.
+        MaxNumberOfStakersExceeded,
+        /// Targets must be operated contracts
+        NotOperatedContract,
+        /// Contract isn't staked.
+        NotStakedContract,
+        /// Unstaking a contract with zero value
+        UnstakingWithNoValue,
+        /// The contract is already registered by other account
+        AlreadyRegisteredContract,
+        /// User attempts to register with address which is not contract
+        ContractIsNotValid,
+        /// Claiming contract with no developer account
+        ContractNotRegistered,
+        // TODO: make use of this?
+        /// Missing deposit for the contract registration
+        InsufficientDeposit,
+        /// This account was already used to register contract
+        AlreadyUsedDeveloperAccount,
+        /// Unexpected state error, used to abort transaction. Used for situations that 'should never happen'.
+        UnexpectedState,
+        /// Report issue on github if this is ever emitted
+        UnknownStartStakingData,
+        /// Report issue on github if this is ever emitted
+        UnknownEraReward,
+        /// There are no funds to reward the contract. Or already claimed in that era
+        NothingToClaim,
+        /// Contract already claimed in this era and reward is distributed
+        AlreadyClaimedInThisEra,
+        /// To register a contract, pre-approval is needed for this address
+        RequiredContractPreApproval,
+        /// Contract is already part of pre-apprved list of contracts
+        AlreadyPreApprovedContract,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(now: BlockNumberFor<T>) -> Weight {
```

Comment:
It could be replace by OnUnbalanced trait implementation, you could handle incoming transfer and account it as incoming balance (in accumulator). Instead of operating of runtime-level information, IMO pallet shouldn’t mind that.